File: frontend/src/services/role.ts
import api from './api';

export interface Role {
  id: number;
  name: string;
  code: string;
  description: string;
  status: string;
  sort: number;
  created_at: string;
  updated_at: string;
  users?: any[];
  menus?: any[];
}

export interface CreateRoleRequest {
  name: string;
  code: string;
  description?: string;
  status?: string;
  sort?: number;
}

export interface UpdateRoleRequest {
  name?: string;
  code?: string;
  description?: string;
  status?: string;
  sort?: number;
}

export const roleService = {
  getRoles: async (): Promise<Role[]> => {
    const response = await api.get('/roles');
    return response.data.roles;
  },

  getRole: async (id: number): Promise<Role> => {
    const response = await api.get(`/roles/${id}`);
    return response.data;
  },

  createRole: async (data: CreateRoleRequest): Promise<Role> => {
    const response = await api.post('/roles', data);
    return response.data.role;
  },

  updateRole: async (id: number, data: UpdateRoleRequest): Promise<Role> => {
    const response = await api.put(`/roles/${id}`, data);
    return response.data.role;
  },

  deleteRole: async (id: number): Promise<void> => {
    await api.delete(`/roles/${id}`);
  },

  assignMenus: async (id: number, menuIds: string[]): Promise<void> => {
    await api.post(`/roles/${id}/menus`, { menu_ids: menuIds });
  },

  assignUsers: async (id: number, userIds: string[]): Promise<void> => {
    await api.post(`/roles/${id}/users`, { user_ids: userIds });
  },
};

File: backend/internal/sys/service/role.go
package service

import (
	"go-admin/internal/sys/model"
	"go-admin/internal/utils"

	"gorm.io/gorm"
)

type RoleService struct {
	db *gorm.DB
}

func NewRoleService(db *gorm.DB) *RoleService {
	return &RoleService{db: db}
}

func (s *RoleService) CreateRole(role *model.Role) error {
	// 生成雪花ID
	role.ID = utils.GenerateID()
	return s.db.Create(role).Error
}

func (s *RoleService) GetRoleByID(id int64) (*model.Role, error) {
	var role model.Role
	err := s.db.Preload("Users").Preload("Menus").First(&role, id).Error
	return &role, err
}

func (s *RoleService) UpdateRole(role *model.Role) error {
	return s.db.Save(role).Error
}

func (s *RoleService) DeleteRole(id int64) error {
	return s.db.Delete(&model.Role{}, id).Error
}

func (s *RoleService) ListRoles() ([]model.Role, error) {
	var roles []model.Role
	err := s.db.Find(&roles).Error
	return roles, err
}

func (s *RoleService) AssignMenus(roleID int64, menuIDs []int64) error {
	var role model.Role
	if err := s.db.First(&role, roleID).Error; err != nil {
		return err
	}

	var menus []model.Menu
	if err := s.db.Where("id IN ?", menuIDs).Find(&menus).Error; err != nil {
		return err
	}

	// 清除现有关联
	if err := s.db.Model(&role).Association("Menus").Clear(); err != nil {
		return err
	}

	// 添加新关联
	if len(menus) > 0 {
		if err := s.db.Model(&role).Association("Menus").Append(menus); err != nil {
			return err
		}
	}

	return nil
}

func (s *RoleService) AssignUsers(roleID int64, userIDs []int64) error {
	var role model.Role
	if err := s.db.First(&role, roleID).Error; err != nil {
		return err
	}

	var users []model.User
	if err := s.db.Where("id IN ?", userIDs).Find(&users).Error; err != nil {
		return err
	}

	// 清除现有关联
	if err := s.db.Model(&role).Association("Users").Clear(); err != nil {
		return err
	}

	// 添加新关联
	if len(users) > 0 {
		if err := s.db.Model(&role).Association("Users").Append(users); err != nil {
			return err
		}
	}

	return nil
}
